Proper public keys implementation

arianvp opened the pull request:
The current webauthn standard is rather ambigious again. So I think
there are some mistakes in the implementation that make our public
key decoder too crypto-agile.

The new draft of the webauthn standard is not ambigious, but also
has an issue I'd like to be clarified first:

w3c/webauthn#1446

it probably means `alg` should imply `crv` and we should remove
support for Ed448 or wait for the spec to be more fleshed out

Fixes https://github.com/arianvp/haskell-fido2/issues/20

arianvp committed:
Make ghcide work on all components

arianvp committed:
Implement proper public key decoding

The current webauthn standard is rather ambigious again. So I think
there are some mistakes in the implementation that make our public
key decoder too crypto-agile.

The new draft of the webauthn standard is not ambigious, but also
has an issue I'd like to be clarified first:

https://github.com/w3c/webauthn/issues/1446

it probably means `alg` should imply `crv` and we should remove
support for Ed448 or wait for the spec to be more fleshed out

arianvp commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+
+data ECDSAKey = ECDSAKey ECDSAIdentifier ECDSA.PublicKey deriving (Eq, Show)
+
+instance Arbitrary ECDSAKey where
```
Probaly move all these arbitrary instances to the test module?

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+    toAlg (-35) = pure $ ECDSAIdentifier ES384
+    toAlg (-36) = pure $ ECDSAIdentifier ES512
+    toAlg (-8) = pure EdDSA
+    toAlg _ = fail "Unsupported `alg`"
```
I double checked these constants (also the encode), they agree with https://www.iana.org/assignments/cose/cose.xhtml#algorithms.

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+      ]
+
+randomKey1 :: Gen Ed25519.PublicKey
+randomKey1 = do
```
Maybe a more descriptive name would be `randomKeyEd25519`. Also for the other `randomKey` functions.

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+-- https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#ctap2-canonical-cbor-encoding-form
+decodePublicKey :: Decoder s PublicKey
+decodePublicKey = do
+  _n <- CBOR.decodeMapLenCanonical
```
If this value is not used, why is it encoded in the first place? Could bad inputs be accepted by ignoring this value?

arianvp replied:
This is useful when reading arbitrary maps. So you can have a loop `n -> 0` and decode keypairs.   The amount of keypairs in our case is completeley static . Either 5 or 6 depending on keytype. And we fail anyway if there is no input left; so I saw no harm in not checking it.

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+  ktyKey <- CBOR.decodeIntCanonical
+  when (ktyKey /= 1) $ fail "Expected required `kty`"
+  kty <- CBOR.decodeIntCanonical
+  case kty of
```
Do I understand correctly that

* `kty` is refers to “key type” in RFC8152, with “key” as in “cryptographic key”.
* We are decoding a CBOR map, which is serialized as key-value pairs, with “key” as in “key-value key”.
* `ktyKey` is the first key (as in key-value key) in the CBOR map.
* `kty` is the value that goes with it.
* RFC8152 demands that there is at least a key-value pair present where the key is the integer 1, and the associated value is the value of the `kty` field.

[CBOR says](https://tools.ietf.org/html/rfc7049#section-3.7):

> The CBOR data model for maps does not allow ascribing semantics to
> the order of the key/value pairs in the map representation.
> Thus, it would be a very bad practice to define a CBOR-based protocol
> in such a way that changing the key/value pair order in a map would
> change the semantics, apart from trivial aspects (cache usage, etc.).
> (A CBOR-based protocol can prescribe a specific order of
> serialization, such as for canonicalization.)

So is it safe to assume that `kty` is the first key-value pair?

arianvp replied:
Yes. the maps must be CTAP2 canonical encoded (as marked in the comment):
https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#ctap2-canonical-cbor-encoding-form

which says keys will be lexicographically sorted by their byte-representation.

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+  kty <- CBOR.decodeIntCanonical
+  case kty of
+    1 -> EdDSAPublicKey <$> decodeEdDSAKey
+    2 -> ECDSAPublicKey <$> decodeECDSAPublicKey
```
From https://tools.ietf.org/html/rfc8152#section-13, I understand that `1` means “Octet Key Pair” and `2` means “Elliptic Curve Keys with x- and y-coordinate pair”.

It was not immediately obvious to me why EdDSA uses the former and ECDSA uses the latter, until I found the table on the next page: https://tools.ietf.org/html/rfc8152#section-13.1

I think it would be good to add a short comment about the meaning of these magic numbers and the relation between `kty` and the algorithm.

Also, the fact that we can tell the algorithm from the `kty` is just a happy coincidence because we only support two algorithms, and one happens to use “octet key pair” and the other happens to use a coordinate, right? How would a decoder do this in general?

arianvp replied:
Correct. it's a happy coincidence.  I thought of making it more explicit  by nesting an additional sum type    `data Key = OKPKey (... edDSA and others) | ECCKey (ECDSA and.. otherS?)`
but removed the indirection later. 

I think a comment is sufficient instead. naming the constants probably helps as well for readability

arianvp replied:
I resolved this by at least having constants for ECC and OKP now

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+    x -> fail $ "unexpected kty: " ++ show x
+
+decodeEdDSAKey :: Decoder s EdDSAKey
+decodeEdDSAKey = do
```
Decoding here also seems to rely on the order of the key-value pairs, is that safe?

arianvp commented:
Specifically [Section 5.8.5.](https://w3c.github.io/webauthn/#sctn-alg-identifier) of the draft forces us to make `alg` uniquely define `kty` and `crv`; which is not how COSE was originally designed. This is all super confusing.

> Keys with algorithm ES256 (-7) MUST specify P-256 (1) as the crv parameter.
> Keys with algorithm ES384 (-35) MUST specify P-384 (2) as the crv parameter.
> Keys with algorithm ES512 (-36) MUST specify P-521 (3) as the crv parameter.
> Keys with algorithm EdDSA (-8) MUST specify Ed25519 (6) as the crv parameter.

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+  when (xKey /= -2) $ fail "Expected required `x`"
+  x <- os2ip <$> CBOR.decodeBytesCanonical
+  yKey <- CBOR.decodeIntCanonical
+  when (yKey /= -3) $ fail "Expected required `x`"
```
s/x/y/

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+  y <- case tokenType of
+    -- TODO(arianvp): Implement compressed curve. Waiting for
+    -- https://github.com/haskell-crypto/cryptonite/issues/302
+    CBOR.TypeBool -> fail "Compressed format not supported _yet_ See Issue number X"
```
Is there a concrete issue number for X?

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+    CBOR.TypeBytes -> os2ip <$> CBOR.decodeBytesCanonical
+    _ -> fail "Unexpected token type"
+  let point = ECC.Point x y
+  when (not (ECC.isPointValid curve point)) $ fail "point not on curve"
```
:+1: for checking this.

(Would have been nicer if the `ECC` only exposed something like `fromCoordinate :: Integer -> Integer -> Maybe ECC.Point`, then you can’t forget to check.)

arianvp replied:
This function exists; but it isn't exported in the library :/ https://github.com/haskell-crypto/cryptonite/blob/master/Crypto/ECC/Simple/Prim.hs#L168

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+  CBOR.encodeMapLen 5
+    <> CBOR.encodeInt 1
+    <> CBOR.encodeInt 2
+    <> CBOR.encodeInt 3
```
There are a lot of magic values here. It would be helpful, both for verification, and for maintainability in case new key types need to be supported, to give them a name in a let-binding.

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+    Right _ -> Nothing
+
+verify :: PublicKey -> ByteString -> ByteString -> Bool
+verify key msg sig =
```
One downside of a `verify` that returns a bool is that you can forget to call it. One way to mitigate this (until we get Linear Haskell :smile:) is to fuse deserialization of the message with verification, something like

```haskell
verifyAndDecode :: PublicKey -> Message ByteString -> Signature -> Maybe DecodedMessage
```
is there a sensible decode operation for the message to do this with?

A different option would be to return some kind of validity token of which the constructor is not exposed, and to have remaining functions take such a token as input.

arianvp replied:
No not really.  `msg =  authenticatorDataBytes || sha256(clientDataJSONBytes)`
so basically involves concatenating two unparsed values, instead of doing calculations on our
parsed datatypes.  There's not much sensible stuff to parse out of it; as at that point we already have the parsed `authenticatorData` and `clientData` anyway

https://github.com/arianvp/haskell-fido2/blob/master/fido/Crypto/Fido2/Assertion.hs#L98

arianvp replied:
Hmm yeh we could still do this...  i'll add this as a todo for later as IDK what is the most ergonomic.

Both attestation and assertion use the same signature format so we have that going for us I think

ruuda commented on fido/Crypto/Fido2/PublicKey.hs:
```diff
@@ -0,0 +1,279 @@
+    2 -> ECDSAPublicKey <$> decodeECDSAPublicKey
+    x -> fail $ "unexpected kty: " ++ show x
+
+decodeEdDSAKey :: Decoder s EdDSAKey
```
All of the other decoders have `PublicKey` in the name, but this one just `Key`. But “key” here does refer to the public key, right? To avoid confusion, I would call this one `decodeEdDSAPublicKey`.

ruuda commented on tests/Spec.hs:
```diff
@@ -53,6 +57,12 @@ main = Hspec.hspec $ do
+      property $ \key -> do
+        let bs = CBOR.toLazyByteString (PublicKey.encodePublicKey key)
+        let rs = snd <$> CBOR.deserialiseFromBytes PublicKey.decodePublicKey bs
+        rs == pure key
```
:+1: Good to have a roundtripping test. Did you check whether this test [covers](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/profiling.html#hpc) all of the encoding and decoding logic?

Also, it would be good to fuzz the decoder too; right now we are only feeding it well-behaved inputs, not arbitrary bytestrings. We can let QuickCheck feed arbitrary byte strings of course, but that is not going to be effective because it is not coverage-driven. I don’t know how to fuzz a Haskell program with coverage-guided fuzzing, I think [AFL in QEMU mode](https://github.com/google/AFL/blob/master/qemu_mode/README.qemu) would work, but I haven’t tried it.

arianvp replied:
I added coverage reports to CI; and yes it's covered for all happy paths.

ruuda commented on server/Database.hs:
```diff
@@ -134,19 +134,21 @@ getCredentialsByUserId (Transaction conn) (UserId (URLEncodedBase64 userId)) = d
   where
-    mkCredential (id, x, y) = do
+    mkCredential (id, publicKey) = do
       -- TODO(#22): Convert to the compressed representation so we don't need
```
This comment is now outdated.

arianvp committed:
Fix build of server

arianvp committed:
Try adding coverage report to github actions

arianvp committed:
Try v2

arianvp committed:
Increase coverage a bit by defining domain objects

Instead of translatign to cryptonite primitives, which is a bit lossy,
we use sum types to model our domain! could remove a never reached error
case because of this

arianvp commented:
Fixes https://github.com/arianvp/haskell-fido2/issues/20

arianvp committed:
Give magic constants names for readability

arianvp committed:
Abstract out verify function

Attestation and assertion have the same signature verification
procedure, so we can reuse this later.

arianvp committed:
Add nix result to gitignore

arianvp committed:
Start working in PublicKey tests

arianvp committed:
Test COSEAlgorithmIdentifier encoding

arianvp committed:
Add more tests

arianvp committed:
Remove Ed448

it's not listed as supported for webauthn.

arianvp committed:
Add failing test for  alg implies crv

arianvp committed:
Fix warnings

arianvp committed:
Implement `alg` implies `crv`

arianvp committed:
Increase test coverage

arianvp committed:
Update cabal2nix

arianvp committed:
100% Coverage for attestation

Property tests for all the failure cases at least

arianvp committed:
Style
